test/system/src/test/java/io/pravega/test/system/ConsumptionBasedRetentionWithMultipleReaderGroupsTest.java

```diff
@@ -391,6 +409,129 @@ public void updateRetentionPolicyForCBRTest() throws Exception {
+        AssertExtensions.assertEventuallyEquals("Truncation did not take place at offset 60.", true, () -> controller.getSegmentsAtTime(
+                        new StreamImpl(scope, stream), 0L).join().values().stream().anyMatch(off -> off == 60),
+                5000, 2 * 60 * 1000L);
+    }
+
+    @Test
+    public void testCBRwithControllerAndSegmentStoreFailover() throws Exception {
+
+        String scope = "testCBR3Scope" + RandomFactory.create().nextInt(Integer.MAX_VALUE);
+        String stream = "testControllerFailOverStream" + RandomFactory.create().nextInt(Integer.MAX_VALUE);
+        String readerGroupNmae = "testControllerFailOverReaderGroup" + RandomFactory.create().nextInt(Integer.MAX_VALUE);
+        scaleAndUpdateControllerURI(3);
+        Futures.getAndHandleExceptions(segmentStoreService.scaleService(2), ExecutionException::new);
+        log.info("Successfully statred 2 instance of segment store service");
+
+        assertTrue("Creating scope", streamManager.createScope(scope));
+        assertTrue("Creating stream", streamManager.createStream(scope, stream, STREAM_CONFIGURATION));
+        @Cleanup
+        ConnectionFactory connectionFactory = new SocketConnectionFactoryImpl(ClientConfig.builder().build());
+        @Cleanup
+        ClientFactoryImpl clientFactory = new ClientFactoryImpl(scope, controller, connectionFactory);
+        @Cleanup
+        EventStreamWriter<String> writer = clientFactory.createEventWriter(stream, new JavaSerializer<>(),
+                EventWriterConfig.builder().build());
+        // Write six events.
+        writingEventsToStream(6, writer, scope, stream);
+        @Cleanup
+        ReaderGroupManager readerGroupManager = ReaderGroupManager.withScope(scope, clientConfig);
+        ReaderGroupConfig readerGroupConfig = getReaderGroupConfig(scope, stream, ReaderGroupConfig.StreamDataRetention.MANUAL_RELEASE_AT_USER_STREAMCUT);
+
+        assertTrue("Reader group is not created", readerGroupManager.createReaderGroup(readerGroupNmae, readerGroupConfig));
+        assertEquals(1, controller.listSubscribers(scope, stream).join().size());
+
+        @Cleanup
+        ReaderGroup readerGroup = readerGroupManager.getReaderGroup(readerGroupNmae);
+        AtomicLong clock = new AtomicLong();
+        @Cleanup
+        EventStreamReader<String> reader = clientFactory.createReader(readerGroupNmae + "-" + 1,
+                readerGroupNmae, new JavaSerializer<>(), readerConfig, clock::get, clock::get);
+        // Read two events with reader.
+        readingEventsFromStream(2, reader);
+
+        log.info("{} generating 1st stream-cuts for {}/{}", readerGroupNmae, scope, stream);
+        Map<Stream, StreamCut> streamCuts = generateStreamCuts(readerGroup, reader, clock);
+
+        log.info("{} updating its retention stream-cut to {}", readerGroupNmae, streamCuts);
+        readerGroup.updateRetentionStreamCut(streamCuts);
+
+        //Controller Failover
```

Comment:
If I am not wrong Issue #5556 says to add test case if controller and SS restarts during test. So instead of scaling down controller and SS instances we need to add test case if there is one instance of Controller and SS and restart during test.
Can we add one more scenario with controller and SS restart?

Reply:
Added the controller and SS restart scenario also.